block/testing.py

```diff
@@ -0,0 +1,111 @@
+from dolfin import MPI
+from urllib.parse import quote_plus
+import atexit
+import inspect
+import numpy as np
+import os
+import pathlib
+import pickle
+
+class RegressionFailure(Exception):
+    pass
+
+def _regr_root():
+    return pathlib.Path(__file__).parent.parent / 'data/regression'
+
+_errs = 0
+def _log_or_raise(msg):
+    if os.environ.get('BLOCK_REGRESSION_ABORT'):
+        raise RegressionFailure(msg)
+    else:
+        print('!', msg)
+        global _errs
+        _errs += 1
+        if _errs == 1:
+            @atexit.register
+            def print_msg():
+                print(f'! {_errs} expected test(s) failed')
+                print(f'  Remove file(s) in {_regr_root()}/ and re-run to store new values as reference')
+
+def check_expected(name, vec, show=False, rtol=1e-6, itol=0.1, prefix=None, expected=None):
+    itol += 1
+    if prefix is None:
+        prefix = pathlib.Path(inspect.stack()[1].filename).stem
+    def _decimate(v):
+        # Pickle not supported for block_vec/dolfin_vec, convert to numpy
+        if hasattr(v, 'get_local'):
+            v = v.get_local()
+        if not isinstance(v, np.ndarray):
+            v = np.concatenate(v)
+        # To save disk&repo space, we decimate the vector and calculate mean+rms
+        # within each chunk. This is quite arbitrary, but is intended to be robust
+        # decimation which still catches most regression scenarios in practice.
+        chunks = np.arange(0, len(v), 10)
+        divisor = np.add.reduceat(np.ones_like(v), chunks)
+        v_mean = np.add.reduceat(v, chunks) / divisor
+        v_rms = np.sqrt(np.add.reduceat(v**2, chunks) / divisor)
+        return (v_mean + v_rms) / 2
+    def _l2(v):
+        if np.isscalar(v):
+            return v
+        elif hasattr(v, 'norm'):
+            return v.norm('l2')
+        else:
+            return np.sqrt(np.sum(v**2)/len(v))
```

Comment:
It seems PETSc does not [normalize](https://petsc.org/release/docs/manualpages/Vec/NORM_2/) by `len(v)`

Reply:
Right, thanks!